### DISCUSSION TITLE
How do I define a C macro for a dependency?

### DISCUSSION THREAD

**Discussion by kwsp** [microsoft/vcpkg#31153] [General]
I need a macro to be defined for one of the dependencies I'm building with VCPKG. Preferrably I can configure this in my project's `vcpkg.json` file so builds are reproducible. 

Specifically, I want to define 

```C++
#define WINVER 0x0601
#define _WIN32_WINNT 0x0601
```

where [`0x0601` is Windows 7](https://learn.microsoft.com/en-us/cpp/porting/modifying-winver-and-win32-winnt). I need this defined so Intel TBB doesn't use a new Windows system call introduced in Windows 8.

https://github.com/oneapi-src/oneTBB/blob/d238e987a53f43f84b41d0c14dbc0089c45e5e65/src/tbb/global_control.cpp#L92-L101

I have to maintain software for a super legacy system running Win 7 and the only thing preventing my builds from work on Windows 7 is this macro.

**Comment by Osyotr** [comment]
Either `set(VCPKG_CXX_FLAGS "/DWINVER=0x0601 /D_WIN32_WINNT=0x0601")` (same for `VCPKG_C_FLAGS`) in your custom triplet or provide a custom toolchain where you can set these definitions.
https://learn.microsoft.com/en-us/vcpkg/users/triplets#vcpkg_cxx_flags

**Reply by dretax** [reply]
@Osyotr can you actually pass arguments somehow to vcpkg.exe or in the vcpkg.json to declare macros?

**Reply by Osyotr** [reply]
Maybe something like this `--cmake-args=-DVCPKG_CXX_FLAGS=\"/DWINVER=0x0601\"` though I would really write a custom triplet.

**Reply by dretax** [reply]
@Osyotr I made my triplet file. Is there a way to use user_settings.h from wolfssl in some way?
Platforms somehow have their own dirs, and i just included WIN/user_settings.h in my own project atleast when i used to compile manually.

**Reply by Osyotr** [reply]
What do you mean "use user_settings.h from wolfssl"?

**Reply by dretax** [reply]
@Osyotr wolfssl declares lots of macros based on many header files.
Basically like including certain settings per platform like WIN,ARDUINO, etc...
In the end i ended up declaring all the windows macros manually...

Only problem I see rn, Is that I'm not getting the x86 wolfssl built when set to x86 platform.

**Comment by kwsp** [comment]
Thanks @Osyotr ! I solved my problem with the following:

Custom triplet file at `<project root>/triplets/x64-windows7.cmake`

```cmake
# From x64-windows.cmake
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE dynamic)

# Define macros for Windows 7.
if (PORT MATCHES "tbb")
  set(VCPKG_CXX_FLAGS "/DWINVER=0x0601 /D_WIN32_WINNT=0x0601")
  set(VCPKG_C_FLAGS "/DWINVER=0x0601 /D_WIN32_WINNT=0x0601")
endif()
```
`vcpkg.json`
```json
{
    ...
    "vcpkg-configuration": {
      "overlay-triplets": ["./triplets"]
    }
}
```

I had to use the `if (PORT MATCHES "tbb")` guard in the triplet because the `/DWINVER` define format only works for MSVC, and would break on my other deps that build with MinGW (e.g. Lapack). Do you have any tricks to make this work for all dependencies that build with MSVC?

**Reply by Neumann-A** [reply]
Put it in the toolchain for MSVC? Also you can use - flags with MSVC so `-DWINVER=0x0601` will also be accepted
